core/templates/dev/head/services/speech-synthesis-chunker.service.spec.ts

```diff
@@ -32,87 +31,199 @@ describe('Speech Synthesis Chunker Service', () => {
+    });
+  });
+
+  describe('convertToSpeakableText', () => {
+    it('should properly convert to speakable text when raw_latex-with-value' +
+      ' attribute is setted', () => {
+      const html = (
+        '<oppia-noninteractive-math raw_latex-with-value="5-1">' +
+        '</oppia-noninteractive-math>' +
+        '<li>Speech</li>' +
+        '<li>Text</li>'
+      );
+
+      expect(speechSynthesisChunkerService.convertToSpeakableText(html))
+        .toBe('5 minus 1Speech. Text. ');
+    });
+
+    it('should properly convert to speakable text when text-with-value' +
+      ' attribute is setted', () => {
+      const html = (
+        '<oppia-noninteractive-link text-with-value="&quot;">' +
+        '</oppia-noninteractive-link>' +
+        '<li>"Speech"</li>' +
+        '<li>Text</li>'
+      );
+
+      expect(speechSynthesisChunkerService.convertToSpeakableText(html))
+        .toBe('Speech. Text. ');
+    });
+  });
+
+  describe('speak', function() {
+    const MockSpeechSynthesisUtteranceConstructor = (
+      SpeechSynthesisUtterance);
+    const mockSpeechSynthesisUtteran = {
+      speak: () => {},
+      onend: () => {}
+    };
+
+    beforeEach(() => {
+      spyOn(window, 'SpeechSynthesisUtterance').and.returnValue(
+        // @ts-ignore
+        mockSpeechSynthesisUtteran);
+    });
+
+    it('should not speak when chunk is too short', () => {
+      const speechSynthesisUtterance = (
+        new MockSpeechSynthesisUtteranceConstructor('a'));
+      const callbackSpy = jasmine.createSpy('callback');
+      speechSynthesisChunkerService.speak(
+        speechSynthesisUtterance, callbackSpy);
+
+      expect(callbackSpy).toHaveBeenCalled();
+    });
+
+    it('should not speak when chunk is a falsy value', () => {
+      const speechSynthesisUtterance = (
+        new MockSpeechSynthesisUtteranceConstructor(''));
+      const callbackSpy = jasmine.createSpy('callback');
+      speechSynthesisChunkerService.speak(
+        speechSynthesisUtterance, callbackSpy);
+
+      expect(callbackSpy).toHaveBeenCalled();
+    });
+
+    it('should speak speech twice', () => {
+      jasmine.clock().uninstall();
+      jasmine.clock().install();
+
+      const speakSpy = spyOn(window.speechSynthesis, 'speak').and
+        .callFake(function() {
+          mockSpeechSynthesisUtteran.onend();
+        });
+      const speechSynthesisUtterance = (
+        new MockSpeechSynthesisUtteranceConstructor(
+          'Value inside utterance for testing purposes.' +
+          ' This is the next chunk'));
+      const callbackSpy = jasmine.createSpy('callback');
+      speechSynthesisChunkerService.speak(
+        speechSynthesisUtterance, callbackSpy);
+
+      jasmine.clock().tick(5000);
+
+      expect(callbackSpy).toHaveBeenCalled();
+      expect(speakSpy).toHaveBeenCalledTimes(2);
+
+      jasmine.clock().uninstall();
+    });
+
+    it('should speak once when cancel is requested', () => {
```

Comment:
I'm not sure what we are testing here, can you please explain?

Reply:
Sure. This block the `SpeechSynthesisUtterance` has two phrases, so `speak` method will speak each phrase separately (two times). However, if `cancel` method is called during speak is being executed, just only the first phrase will be spoken.